include/swift/SIL/GraphOperationInfo.h

```diff
@@ -69,11 +69,18 @@ struct GraphOperationInfo {
     UnknownShapeListAttribute,
 
     /// A metatype of a TensorFlow value type or aggregate of TensorFlow value
-    /// types that should be lowered into a list of type attributes.
+    /// types. Deabstraction lowers this to TFDataTypeAttribute.
     ///
     /// Written as named argument with "$typeList" suffix.
     TypeListAttribute,
 
+    /// An integer, or an aggregate containing a single integer, representing a
```

Comment:
do we want to remove the "or an aggregate containing a single integer" case by calling `lookThroughSingleElementAggregates()` in the producer of this instance, so that consumer code will be kept simple?

Reply:
This starts out as a not-necessarily-constant SILValue, so I'd have to do this extraction in SILGen (or maybe CSApply?) in order to remove "or an aggregate containing a single integer". This would be a bit difficult because those early phases don't currently know anything about the different kinds of attributes and they don't currently do any transformations of them.

An easier change that removes a case would be to make it always be `TensorDataType`. (Specifically, `convertSwiftTypeToTFSymbolicValue` can make an aggregate SymbolicValue instead of an integer SymbolicValue).

That would be kind of wasteful because lowering will just unwrap it again, and also it complicates `convertSwiftTypeToTFSymbolicValue`. But maybe the simplification of this data structure is worth it?

I'll leave it as two cases for now, but let me know if you think I should really make it one case.

Reply:
Thanks for the context. Given making it one case does not reduce overall complexity, having the current two-case design makes sense. Consider adding an impl-related code comment here to briefly capture the rationale?